src/lib/mina/zkapp.ts

```diff
@@ -157,7 +159,20 @@ function method<K extends string, T extends SmartContract>(
     ZkappClass._maxProofsVerified,
     methodEntry.proofArgs.length
   ) as 0 | 1 | 2;
-  let func = descriptor.value as AsyncFunction;
+  let func = function (this: SmartContract, ...args) {
+    const callsCount = Field(
+      args
+        .filter(
+          (v: unknown): v is ZkProgram<any, any> => v instanceof ZkProgram
+        )
+        .reduce(
+          (acc: number, cur: ZkProgram<any, any>) => acc + cur.getCallsCount(),
+          0
+        )
+    );
+    this.emitEvent(MetadataEvent.key, MetadataEvent.fromValue({ callsCount }));
+    return descriptor.value(...args);
+  } as AsyncFunction;
```

Comment:
* this seems like the wrong place to add wrapper logic that runs on every smart contract call
  * there's already `wrapMethod()` for that, called on the next line - why not add your logic there?
* just to call it out: if you add logic that runs on every smart contract call and influences the circuit and the shape of the transaction, that's a breaking change
* the current logic for finding recursive proof arguments seems invalid, because these proof arguments are not instances of ZkProgram -- ZkProgram isn't even a class